pkg/controller/connectivitycheck/connectivity_check_controller.go

```diff
@@ -0,0 +1,403 @@
+}
+
+func (c *connectivityCheckTemplateProvider) getTemplatesForGenericPodServiceEndpointsChecks(recorder events.Recorder) []*v1alpha1.PodNetworkConnectivityCheck {
+	var templates []*v1alpha1.PodNetworkConnectivityCheck
+	addresses, err := c.listAddressesForGenericPodServiceEndpoints(recorder)
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "unable to determine openshift-network-diagnostics network-check-target endpoints: %v", err)
+		return nil
+	}
+
+	for _, address := range addresses {
+		templates = append(templates, NewPodNetworkConnectivityCheckTemplate(net.JoinHostPort(address.hostName, address.port), "openshift-network-diagnostics", withTarget("network-check-target", address.nodeName)))
+	}
+	return templates
+}
+
+// listAddressesForGenericPodServiceEndpoints returns network-check-target service endpoints ip
+func (c *connectivityCheckTemplateProvider) listAddressesForGenericPodServiceEndpoints(recorder events.Recorder) ([]endpointInfo, error) {
+	var results []endpointInfo
+	endpoints, err := c.diagnosticsEndpointsLister.Endpoints("openshift-network-diagnostics").Get("network-check-target")
+	if err != nil {
+		return nil, err
+	}
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			for _, port := range subset.Ports {
+				results = append(results, endpointInfo{
+					hostName: address.IP,
+					port:     strconv.Itoa(int(port.Port)),
+					nodeName: *address.NodeName,
+				})
+			}
+		}
+	}
+	return results, nil
+}
+
+func (c *connectivityCheckTemplateProvider) getTemplatesForAPILoadBalancerChecks(recorder events.Recorder) []*v1alpha1.PodNetworkConnectivityCheck {
+	var templates []*v1alpha1.PodNetworkConnectivityCheck
+	infrastructure, err := c.infrastructureLister.Get("cluster")
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "error detecting api load balancer endpoints: %v", err)
+		return nil
+	}
+
+	apiUrl, err := url.Parse(infrastructure.Status.APIServerURL)
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "error detecting external api load balancer endpoint: %v", err)
+
+	} else {
+		templates = append(templates, NewPodNetworkConnectivityCheckTemplate(apiUrl.Host, "openshift-network-diagnostics", withTarget("load-balancer", "api-external")))
+	}
+
+	apiInternalUrl, err := url.Parse(infrastructure.Status.APIServerInternalURL)
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "error detecting internal api load balancer endpoint: %v", err)
+	} else {
+		templates = append(templates, NewPodNetworkConnectivityCheckTemplate(apiInternalUrl.Host, "openshift-network-diagnostics", withTarget("load-balancer", "api-internal")))
+	}
+	return templates
+}
+
+type endpointInfo struct {
+	hostName string
+	port     string
+	nodeName string
+}
+
+func withTarget(label, nodeName string) func(check *v1alpha1.PodNetworkConnectivityCheck) {
```

Comment:
Why?

Reply:
Because this is not calling `withTarget` with a label and a node name:

https://github.com/openshift/cluster-network-operator/blob/0542083f9f705c934096215413acc25202ce0650/pkg/controller/connectivitycheck/connectivity_check_controller.go#L220

nor is this:

https://github.com/openshift/cluster-network-operator/blob/0542083f9f705c934096215413acc25202ce0650/pkg/controller/connectivitycheck/connectivity_check_controller.go#L333

and most (almost all) places where it is being called.

Reply:
well, it is here:

https://github.com/openshift/cluster-network-operator/blob/0542083f9f705c934096215413acc25202ce0650/pkg/controller/connectivitycheck/connectivity_check_controller.go#L184

The reason why in those lines you referred are not nodes is because they don't apply to a node as they are services and LBs respectively.

So I take when you mean 'label and nodeName' you really mean nodeName?